flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationType.scala

```diff
@@ -0,0 +1,171 @@
+      loss: Double,
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regularizationParameter: Double) : Double ={
+    val adjustedLoss = regLoss(loss, weightVector, regularizationParameter)
+    regGradient(weightVector, lossGradient, regularizationParameter)
+
+    adjustedLoss
+  }
+
+  /** Adds regularization gradient to the loss gradient. The gradient is updated in place **/
+  def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regularizationParameter: Double)
+}
+
+/** Performs no regularization, equivalent to $R(w) = 0$ **/
+class NoRegularization extends RegularizationType {
+  /** Adds regularization to the loss value **/
+  override def regLoss(oldLoss: Double, weightVector: FlinkVector, regularizationParameter: Double):
+  Double = {oldLoss}
+}
+
+/** $L_2$ regularization penalty.
+  *
+  * Penalizes large weights, favoring solutions with more small weights rather than few large ones.
+  *
+  */
+class L2Regularization extends DiffRegularizationType {
+
+  /** Adds regularization to the loss value **/
+  override def regLoss(oldLoss: Double, weightVector: FlinkVector, regParameter: Double)
+    : Double = {
+    val brzVector = weightVector.asBreeze
+    oldLoss + regParameter * (brzVector dot brzVector) / 2
+  }
+
+  /** Adds regularization gradient to the loss gradient. The gradient is updated in place **/
+  override def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regParameter: Double): Unit = {
+    BLAS.axpy(regParameter, weightVector, lossGradient)
+  }
+}
+
+/** $L_1$ regularization penalty.
+  *
+  * The $L_1$ penalty can be used to drive a number of the solution coefficients to 0, thereby
+  * producing sparse solutions.
+  *
+  */
+class L1Regularization extends RegularizationType {
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * Implementation was taken from the Apache Spark Mllib library:
+    * http://git.io/vfZIT
+    * @param oldWeights The old weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The current regularization parameter
+    * @return A tuple whose first element is the updated weight FlinkVector and the second is the
+    *         regularization value
+    */
+  override def takeStep(
+      oldWeights: FlinkVector,
+      gradient: FlinkVector,
+      effectiveStepSize: Double,
+      regParameter: Double) {
+    BLAS.axpy(-effectiveStepSize, gradient, oldWeights)
+    val brzWeights = oldWeights.asBreeze.toDenseVector
+
+    // Apply proximal operator (soft thresholding)
+    val shrinkageVal = regParameter * effectiveStepSize
+    var i = 0
+    while (i < brzWeights.length) {
+      val wi = brzWeights(i)
+      brzWeights(i) = signum(wi) * BreezeMax(0.0, abs(wi) - shrinkageVal)
+      i += 1
+    }
+
+    BLAS.copy(brzWeights.fromBreeze, oldWeights)
```

Comment:
Why using for some operations Breeze and for others the BLAS object? With Breeze you should be able to do at least the same as with the BLAS object, right? Wouldn't it make sense to do all with either Breeze or BLAS? That way we would only have one "dependency".